enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,509 @@
+   `k8s.ovn.org/egress-assignable`
+2. User creates a `egressips.k8s.ovn.org` CR with its requested egress IP
+   assignments.
+3. ovnkube-master, having initialized all node level assignment data: performs
+   an assignment and issues a cloud egress IP assignment request by creating a
+   `cloudegressips.network.openshift.io` CR with `.spec.requests` filled out.
+4. Watches for the update of the CR and inspects `.status.assignments` to
+   retrieve the assignment status.
+5. Performs cluster level networking setup for the egress IPs.  
+6. Updates the `egressips.k8s.ovn.org` CR created by the user in 2. with its
+   `.status.assignments`, notifying the user of the final assignment status.
+
+###### openshift-sdn
+
+Note: openshift-sdn does not flag successful or unsuccessful assignment of
+egress IPs in any explicit way, it only logs unsuccessful assignment in its own
+logs. The user takes care of modifying the `hostsubnet` and `netnamespace` CRs,
+but if an incorrect setup is performed (for example by not aligning the egress
+IP with the nodes' subnet range), the user is never notified of this but needs
+to take an active action and look in the openshift-sdn logs.
+
+**Initialization**
+
+1. When openshift-sdn starts: it initializes the `hostsubnet` CR and fills in
+   the fields of each node's subnet information.  
+2. It will then continue watching nodes as to retrieve all nodes'
+   `network.openshift.io/node-cloud-ifaddr` (which will be updated by the cloud
+   component during its initialization) and finalize the initialization of its
+   required assignment data.
+
+**Egress IP assignment**
+
+1. User patches `hostsubnet.egressIP` / `hostsubnet.egressCIDRs`  corresponding
+   to "flagging" the nodes wished to be used for egress assignment
+2. User patches `netnamespace.egressIPs` on the namespace wished to be matched
+   for egress IP assignment
+3. openshift-sdn, having initialized all node level assignment data: performs
+   an assignment and issues a cloud egress IP assignment request by creating a
+   `cloudegressips.network.openshift.io` CR with `.spec.requests` filled out.
+4. Watches for the update of the CR and inspects `.status.assignments` to
+   retrieve the assignment status.
+5. Performs cluster level networking setup for the egress IPs.  
+6. Triggers an `event` to the user notifying successful/unsuccessful assignment.
+
+### Risks and Mitigations
+
+#### Risks
+
+Finalizing the work for 4.8 has a risk arising mainly CI/CD setup for this new
+component. It is imperative that a finalized name is chosen as soon as possible
+as to avoid additional work updating any CI/CD dependencies, post their
+creation.
+
+Given future OpenShift features requiring cloud level modifications of the
+network fabric (once such existing future enhancement is: node port range
+modifications): the proposal here is to call the repository;
+openshift/cloud-network-controller. The name would explicitly indicate that this
```

Comment:
@danwinship / @abhat / @tssurya / @squeed : given the urgency to get this done and done once: I've taken the liberty and propose that the repo be called `cloud-network-controller`, please view this section as to why. Let's have the discussion here, if you don't agree with this.

Reply:
Also, following this it might be better if the CRD's `Kind` is changed from `CloudEgressIP` to something like `CloudPrivateIPAssignment` as to avoid making any reference to other OpenShift features and solely focus on specific "cloud network manipulations".

Reply:
I haven't yet gone through the recent version of the enhancement, will try to do that asap.

Anyways regarding the name, `cloud-network-controller` is surely better than have anything egressIP specific. Same goes for the CRD, imho. I am ok with `CloudPrivateIPAssignment`.

However the term "controller" in my head sounds like we are controlling the network settings in the cloud which is not entirely true since this component just sends creation requests to the cloud since in the end we can't bypass what the cloud thinks - this feature is actually to circumvent the subnet control that cloud has. But that's just me.

Maybe the goal of this component for further enhancements might make the name apt, dunno. I don't have a better name in mind - some that just popped into mind were `cloud-network-resource-allocator/requestor/generator/creator`, `cloud-network-operator`. I am not sure if they sound any better outside my head heh.

So I'm ok with the name if others feel fine with it. Besides, I don't want us to spend anymore cycles for the naming part.

Reply:
I thought about `cloud-network-operator`, but since this component wouldn't really be ["packaging, deploying, and managing a Kubernetes application"](https://www.redhat.com/en/topics/containers/what-is-a-kubernetes-operator) I refrained from it. I thought controller seems descriptive enough since it will be running a control loop and perform operations accordingly, as Kubernetes/OpenShift controllers do.

Reply:
`cloud-network-controller` is too generic a name. I think we had talked about `cloud-network-config-controller` but @danwinship had some hesitation to call it something that generic as well given the security groups CRD support is still not work that is slated.

Reply:
Personally, I would like for it to not represent something too specific.

Reply:
Dan, said he's fine with the name `cloud-network-controller` (https://coreos.slack.com/archives/CDCP2LA9L/p1612457607380800) so I think we can call it either. If you prefer `cloud-network-config-controller` we can call it that. 

I will start the process on Monday, if there's no objection until then.

Reply:
Ok for the CRD, I think `CloudPrivateIPConfig` makes more sense? Since assignment has a different connotation.

Reply:
I agree - it would also be more in line with the repo name. I will update the PR tomorrow.